src/ArrayComparator.php

```diff
@@ -124,4 +126,19 @@ private function indent(string $lines): string
     {
         return trim(str_replace("\n", "\n    ", $lines));
     }
+
+    private function canonicalize(array &$array): void
+    {
+        if (array_is_list($array)) {
+            sort($array);
+        } else {
+            ksort($array);
+        }
+
+        foreach ($array as &$element) {
+            if (is_array($element)) {
```

Comment:
If an element of the parent $array is an array, won't it get passed to the ArrayComparator@assertEquals() method? It seems like we wouldn't need to recursively sort here if that's true.

Reply:
Recursion is needed. After all, nesting can be as big as you want, for example:
```php
$a = [
    'k' => [
        'h' => [
            'n' => [
                ....
            ],
        ],
    ],
];
```

Reply:
Yes, but `assertEquals()` is going to loop through each sub-element of the array and then call `Comparator@assertEquals()` on that.

If the sub-element is an array, that means it's going to get passed to `ArrayComparator@canonicalize()` again.

So you're sorting everything, and then sorting it again after creating the factory on lines 70-71.

Do I have that wrong?

Reply:
I finally got it. It's brilliant!